streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregate.java

```diff
@@ -0,0 +1,428 @@
+                    final TimeWindow window = new TimeWindow(rightWinStart, rightWinStart + windows.timeDifference);
+                    final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(initializer.apply(), timestamp);
+                    putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+                }
+            }
+
+            //create left window for new record
+            if (!leftWinExists) {
+                final Agg aggValue;
+                final long newTimestamp;
+                if (latestLeftTypeWindow != null) {
+                    aggValue = aggregator.apply(key, value, getValueOrNull(leftWinAgg));
+                    newTimestamp = leftWinAgg.timestamp();
+                } else {
+                    aggValue = aggregator.apply(key, value, initializer.apply());
+                    newTimestamp = timestamp;
+                }
+                final TimeWindow window = new TimeWindow(timestamp - windows.timeDifference, timestamp);
+                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(aggValue, Math.max(timestamp, newTimestamp));
+                putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+            }
+            //create right window for new record
+            if (!rightWindowExists && (foundLeftFirst || prevRightWinAlreadyCreated)) {
+                final TimeWindow window = new TimeWindow(timestamp + 1, timestamp + 1 + windows.timeDifference);
+                final ValueAndTimestamp<Agg> valueAndTime = ValueAndTimestamp.make(getValueOrNull(rightWinAgg), Math.max(rightWinAgg.timestamp(), timestamp));
+                putAndForward(window, valueAndTime, key, value, closeTime, timestamp);
+            }
+
+        }
+
+        private boolean isLeftWindow(final KeyValue<Windowed<K>, ValueAndTimestamp<Agg>> window) {
+            return window.key.window().end() == window.value.timestamp();
+        }
+
+        private void putAndForward(final Window window, final ValueAndTimestamp<Agg> valueAndTime, final K key,
+                                   final V value, final long closeTime, final long timestamp) {
+            final long windowStart = window.start();
+            final long windowEnd = window.end();
+            if (windowEnd > closeTime) {
+                //get aggregate from existing window
+                final Agg oldAgg = getValueOrNull(valueAndTime);
+                //add record's value to existing aggregate
+                final Agg newAgg = aggregator.apply(key, value, oldAgg);
+
+                windowStore.put(key,
+                        ValueAndTimestamp.make(newAgg, Math.max(timestamp, valueAndTime.timestamp())),
+                        windowStart);
+                tupleForwarder.maybeForward(
+                        new Windowed<K>(key, window),
+                        newAgg,
+                        sendOldValues ? oldAgg : null,
+                        windowStart);
```

Comment:
I think we need to pass in the new maximum window timestamp here, not the window start time

Reply:
good catch